kong/db/schema/entities/filter_chains.lua

```diff
@@ -31,6 +34,36 @@ local filter = {
     { config  = { type = "string", required = false, }, },
     { enabled = { type = "boolean", default = true, required = true, }, },
   },
+  entity_checks = {
+    { custom_entity_check = {
+        field_sources = { "name", "config" },
+        fn = function(entity)
+          kong = kong or _G.kong
+          local filter, err = kong.db.filters:select_by_name(entity.name)
+          if err then
+            error(err)
+
+          elseif not filter or not filter.config_schema then
+            return true
+          end
+
+          local user_schema
+          user_schema, err = kong.db.schemas:select(filter.config_schema)
+
+          if user_schema then
+            return schema.validate_entity(user_schema, entity.config)
+
+          elseif err then
+            error(err)
+
+          else
+            -- this *should* be unreachable
+            error("schema for proxy-wasm filter '" .. entity.name .. "' not found")
+          end
+        end,
```

Comment:
This validation function is still in the "just make the tests pass" stage because I didn't want to tinker with it endlessly. The performance outlook of this is not good, especially given that this is per-filter validation, so it will execute in a loop (once for each of `filter_chains[].filters`).

Seeking some pointers on how to best optimize this (best referring to the maintainability moreso than the performance). Should I just use `kong.cache:get()` to memoize `select()` ops? Should I try to add a one-off JOIN query to the DAO so that we can reduce these two `select()`s into one?

None of the other code in `kong/db/*` seems to depend upon `kong.cache`, so it feels like it might be bad practice to reach for memoization here.

Alternatively, maybe it's fine to leave as is for the time being? This is a brand new feature _within_ another brand new feature (Wasm) which is still in tech preview, so I am open to the idea that it doesn't need to be fully-optimized yet.

Reply:
I think something like that would be nice at wiping out all the lookup logic from this function. Something along the lines of...

```lua
local filter = {
  type = "record",
  fields = {
    { name    = { type = "string", required = true, one_of = wasm.filter_names,
                  err = "no such filter", }, },

    { config  = { type = "string", required = false, 
                  schema = "filters://$self.name/config_schema",
                },
    },

    --[[ other fields omitted for brevity ]]--
  },
}
```

I'm slightly unsure about pursuing this because the multiple levels of indirection mean that we _are_ asking a lot of the code that needs to resolve this reference (in order for it to be generic enough to be a native feature of the schema library). We would essentially be solving for the most complicated use-case right out of the gate.

---